jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,246 @@
+    (strong version of second one).
+    Args:
+        func: function to minimise, should take a single flat parameter, similar to numpy
+        x0: initial guess of parameter
+        max_iterations: int or None which means inf
+        analytic_initial_hessian: whether to use JAX to get hessian for the initialisation, otherwise it's eye.
+        g_tol: float, stopping criteria is |grad(func)|_2 \lt g_tol
+
+    Returns: BFGSResults
+
+    """
+    state = BFGSResults(converged=False,
+                        failed=False,
+                        k=0,
+                        nfev=0,
+                        ngev=0,
+                        nhev=0,
+                        x_k=x0,
+                        f_k=None,
+                        g_k=None,
+                        H_k=None)
+
+    if max_iterations is None:
+        max_iterations = np.inf
+
+    D = x0.shape[0]
+
+    if analytic_initial_hessian:
+        hess = hessian(func, argnums=0)
+        initial_B = hess(x0)
+        # TODO: pinc may give pathological behaviour if function not C^2 and hess is all zeros
+        initial_H = np.linalg.pinv(initial_B)
+        state = state._replace(nhev=state.nhev + 1)
+    else:
+        initial_H = np.eye(D)
+
+    value_and_grad = jax.value_and_grad(func)
+
+    f_0, g_0 = value_and_grad(x0)
+    state = state._replace(f_k=f_0, g_k=g_0, H_k=initial_H, nfev=state.nfev + 1, ngev=state.ngev + 1,
+                           converged=np.linalg.norm(g_0) < g_tol)
+
+    while not state.converged and not state.failed and state.k < max_iterations:
+        p_k = -np.dot(state.H_k, state.g_k)
+        line_search_results = line_search_nojit(value_and_grad, state.x_k, p_k, f_0=state.f_k, g_0=state.g_k)
+        state = state._replace(nfev=state.nfev + line_search_results.nfev,
+                               ngev=state.ngev + line_search_results.ngev,
+                               failed=line_search_results.failed)
+        s_k = line_search_results.a_k * p_k
+        x_kp1 = state.x_k + s_k
+        f_kp1 = line_search_results.f_k
+        g_kp1 = line_search_results.g_k
+        # print(g_kp1)
+        # f_kp1, g_kp1 = value_and_grad(x_kp1)
+        # print(g_kp1)
+        y_k = g_kp1 - state.g_k
+        rho_k = np.reciprocal(np.dot(y_k, s_k))
+
+        sy_k = s_k[:, None] * y_k[None, :]
+        w = np.eye(D) - rho_k * sy_k
+        H_kp1 = np.dot(np.dot(w, state.H_k), w.T) + rho_k * s_k[:, None] * s_k[None, :]
+
+        converged = np.linalg.norm(g_kp1) < g_tol
+
+        state = state._replace(converged=converged,
+                               k=state.k + 1,
+                               x_k=x_kp1,
+                               f_k=f_kp1,
+                               g_k=g_kp1,
+                               H_k=H_kp1
+                               )
+        print(state)
+    return state
+
+
+def minimize(func, x0, max_iterations: int = None, analytic_initial_hessian: bool = True, g_tol: float = 1e-8):
```

Comment:
In JAX, we try to copy the interface of numpy/scipy *exactly* for every parameter that matches, unless we have a very good reason otherwise.

In this case, I would suggest copying the exact variable names from SciPy, with all relevant parameters e.g., `minimize(fun, x0, *, method=None, tol=None, options=None)`

`maxiter`, `analytic_initial_hessian` and `gtol` are solver specific options, so they should go into the `options` dict. (It's OK that `analytic_initial_hessian` is something new, not found in SciPy, but the other names should match.)

I am conflicted about whether we should have a default optimization method for `minimize`. Defaulting to BFGS would match SciPy, but I think this behavior is arguably a mistake, because BFGS isn't suitable for high dimensional problems. A better default for SciPy would probably have been L-BFGS. I guess my inclination would be to not include a default optimization method and force the user to explicitly choose, but I'm curious what others think.

Reply:
Done. Also, no default method.